Fix unbind_copy and add its decomposition

rec opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134319

* Fixes https://github.com/pytorch/pytorch/issues/130829

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @XilunWu

rec committed:
Fix and add decomposition for unbind_copy

* Fixes https://github.com/pytorch/pytorch/issues/130829

[ghstack-poisoned]

rec commented:
I had to add two special cases to `out_wrapper` in torch/_prims_common/wrappers.py because `unbind_copy` differs from many other `out=` functions, see the above issue.

rec committed:
Update on "Fix and add decomposition for unbind_copy"


* Fixes https://github.com/pytorch/pytorch/issues/130829

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

amjames commented on torch/_prims/context.py:
```diff
@@ -129,6 +129,8 @@ def __torch_function__(
         elif func is None and isinstance(orig_func, torch._ops.OpOverloadPacket):
             default = getattr(orig_func, "default", None)
+            if default is None:
+                default = getattr(orig_func, orig_func._dir[0], None)
```
This needs to be guarded since there isn't any guarantee that `orig_func._dir` is not empty. If that access is out of bounds then default should remain `None`. As written it would raise an exception.

rec replied:
Urg, fixed, good catch.

amjames commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -19491,6 +19491,27 @@ def sample_inputs_alias_copy(op_info, device, dtype, requires_grad, **kwargs):
+           check_batched_grad=False,
+           skips=(
+               # grad actually does work with out=
+               DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_out_requires_grad_error'),
```
Is there any other example where requires_grad + out= work together? I don't know of any which is making me suspicious about this.

rec replied:
```
from torch.testing._internal.common_methods_invocations import op_db
a = [o for o in op_db if o.supports_out and o.supports_autograd]
print(len(a))  # prints 268
print(len([i for i in a if i.name.endswith('_copy')]))  # prints 10
```

amjames replied:
Those ops have an `out=` overload and support gradients, yes. I am asking if any of those ops also have this test x-failed. The test asserts that an error is raised if we are tracking gradients on the `out=` argument which is expected why is this op an exception to that rule?

rec committed:
Update on "Fix and add decomposition for unbind_copy"


* Fixes https://github.com/pytorch/pytorch/issues/130829

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

rec committed:
Update on "Fix and add decomposition for unbind_copy"


* Fixes https://github.com/pytorch/pytorch/issues/130829

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

amjames commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -4055,31 +4055,34 @@ void split_copy_Tensor_out(const at::Tensor & self, int64_t split_size, int64_t
+// void unbind_copy_int_out(const at::Tensor & self, int64_t dim, at::TensorList  out) {
+//   auto tmp = self.unbind(dim);
+//   copy_tensor_array_to_out("unbind_copy_int_out()", tmp, out);
+// }
```
You can delete this

amjames commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -14639,11 +14639,13 @@
+#   variants: function
+#   dispatch:
+#     CompositeExplicitAutograd: unbind_copy_int_out
+#   tags: view_copy
```
This can be deleted

amjames commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -4055,31 +4055,34 @@ void split_copy_Tensor_out(const at::Tensor & self, int64_t split_size, int64_t
-        "Expected out tensor to have device ", tmp[i].device(), ", but got ", out[i].device(), " instead");
-    out[i].copy_(tmp[i]);
+    TORCH_CHECK(out[i].dtype() == array[i].dtype(),
+        "Expected out tensor to have dtype ", array[i].dtype(), ", but got ", out[i].dtype(), " instead");
```
NIT you can put the index in the error message to indicate which in/out pair had a mismatch. This will also help clarify as the message reads like there is a single tensor input/output. 

dito for device check below

rec replied:
This code is now all automatically generated...

amjames reviewed (commented):
Couple of comments

rec commented:
Oops, sorry, this isn't quite ready for review yet, I'll ping you when it's
all done... it's still in draft!

On Thu, Aug 29, 2024 at 7:48 PM Andrew James ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> Couple of comments
> ------------------------------
>
> In aten/src/ATen/native/TensorShape.cpp
> <https://github.com/pytorch/pytorch/pull/134319#discussion_r1736810751>:
>
> >  }
>
> +// void unbind_copy_int_out(const at::Tensor & self, int64_t dim, at::TensorList  out) {
> +//   auto tmp = self.unbind(dim);
> +//   copy_tensor_array_to_out("unbind_copy_int_out()", tmp, out);
> +// }
>
> You can delete this
> ------------------------------
>
> In aten/src/ATen/native/native_functions.yaml
> <https://github.com/pytorch/pytorch/pull/134319#discussion_r1736811047>:
>
> >
> -- func: unbind_copy.int_out(Tensor self, int dim=0, *, Tensor(a!)[] out) -> ()
> -  variants: function
> -  dispatch:
> -    CompositeExplicitAutograd: unbind_copy_int_out
> +# - func: unbind_copy.int_out(Tensor self, int dim=0, *, Tensor(a!)[] out) -> ()
> +#   variants: function
> +#   dispatch:
> +#     CompositeExplicitAutograd: unbind_copy_int_out
> +#   tags: view_copy
>
> This can be deleted
> ------------------------------
>
> In aten/src/ATen/native/TensorShape.cpp
> <https://github.com/pytorch/pytorch/pull/134319#discussion_r1736815055>:
>
> >      }
> -    TORCH_CHECK(out[i].dtype() == tmp[i].dtype(),
> -        "Expected out tensor to have dtype ", tmp[i].dtype(), ", but got ", out[i].dtype(), " instead");
> -    TORCH_CHECK(out[i].device() == tmp[i].device(),
> -        "Expected out tensor to have device ", tmp[i].device(), ", but got ", out[i].device(), " instead");
> -    out[i].copy_(tmp[i]);
> +    TORCH_CHECK(out[i].dtype() == array[i].dtype(),
> +        "Expected out tensor to have dtype ", array[i].dtype(), ", but got ", out[i].dtype(), " instead");
>
> NIT you can put the index in the error message to indicate which in/out
> pair had a mismatch. This will also help clarify as the message reads like
> there is a single tensor input/output.
>
> dito for device check below
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/134319#pullrequestreview-2269653757>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAB53MUGJIRI6QG633SJ2SLZT5NGDAVCNFSM6AAAAABM7XTVL2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDENRZGY2TGNZVG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
     /t

PGP Key: ***@***.***
*https://tom.ritchford.com <https://tom.ritchford.com>*
*https://tom.swirly.com <https://tom.swirly.com>*

rec committed:
Update on "Fix and add decomposition for unbind_copy"


* Fixes https://github.com/pytorch/pytorch/issues/130829

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

rec committed:
Update on "Fix and add decomposition for unbind_copy"


* Fixes https://github.com/pytorch/pytorch/issues/130829

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

rec commented:
All right, this seems to have completed fine: there's an "unrelated error" which I just re-ran but that seems to be irrelevant.

rec committed:
Update on "Fix and add decomposition for unbind_copy"


* Fixes https://github.com/pytorch/pytorch/issues/130829

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

amjames commented on test/test_ops.py:
```diff
@@ -1070,17 +1074,31 @@ def set_requires_grad(x):
+                op(sample.input, *sample.args, **sample.kwargs, out=out)
+
+        else:
+            result = op(sample.input, *sample.args, **sample.kwargs, out=out)
+            self.assertEqual(result, None)
+            self.assertIsInstance(out, tuple)
+            self.assertNotEqual(out, ())
+
+            [o.backward() for o in out]
+            self.assertTrue(all(o.grad is not None for o in out))
+
+            if op.name == "unbind_copy":  # Only such op right now
+                self.assertEqual({o.shape for o in out}, {out[0].shape})
+                self.assertTrue(all(torch.all(o.grad == 1.0) for o in out))
```
This does not cover many interesting cases for derivative support, so code testing this situation would be best added to `test_ops_gradients.py`. Doing so should allow you to also check the complex dtypes as inputs for backward will be generated correctly. Here you are relying on the coincidental fact that the first sample for `unbind_copy` produces scalar outputs. 

However, getting the out overload to work with gradcheck will be difficult, if not impossible, that kind of reinforces my suspicions that this should not support gradients at all.

rec replied:
All right, I went back to the previous state without autogenerating the code, and made sure by hand that we're emitting the right message.

rec committed:
Update on "Fix and add decomposition for unbind_copy"


* Fixes https://github.com/pytorch/pytorch/issues/130829

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

rec commented:
Should be ready to go!

rec committed:
Update on "Fix and add decomposition for unbind_copy"


* Fixes https://github.com/pytorch/pytorch/issues/130829

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

rec committed:
Update on "Fix unbind_copy and add its decomposition"


* Fixes https://github.com/pytorch/pytorch/issues/130829

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

rec committed:
Update on "Fix unbind_copy and add its decomposition"


* Fixes https://github.com/pytorch/pytorch/issues/130829

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

rec committed:
Update

[ghstack-poisoned]

rec committed:
Update

[ghstack-poisoned]

eellison reviewed (commented):
test failures.. please re-request review when ready

rec committed:
Update

[ghstack-poisoned]

rec commented:
@eellison: so sorry, this was a breakage of several decompositions which I addressed [here](https://github.com/pytorch/pytorch/pull/136653) and was unrelated to this specific pull request.

Now I rebased this pull against that fix, the one CI test breakage above is not related and appears on several other unrelated pull requests.

Have another look?  🙂

rec committed:
Update

[ghstack-poisoned]

rec committed:
Update

[ghstack-poisoned]

rec committed:
Update

[ghstack-poisoned]

rec commented:
@eellison: I rebased, all tests are green and shiny now. 🙂

eellison reviewed (approved):
nice!

eellison commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot revert -m "breaking some executorch tests D64568664" -c ghfirst

The test it fails is actually in the executorch repo and you can see it here https://github.com/pytorch/executorch/blob/7493aaed4477ae302431e31d269312e6de480c7d/kernels/test/op_unbind_copy_test.cpp#L29

```
Note: Google Test filter = OpUnbindCopyIntOutTest.UnbindFailsWithWronglyAllocatedOutput
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from OpUnbindCopyIntOutTest
[ RUN      ] OpUnbindCopyIntOutTest.UnbindFailsWithWronglyAllocatedOutput
.../executorch/kernels/test/op_unbind_copy_test.cpp:292: Failure
Expected: op_unbind_copy_int_out(input, 1, out) throws an exception.
  Actual: it doesn't.

.../executorch/kernels/test/op_unbind_copy_test.cpp:300: Failure
Expected: op_unbind_copy_int_out(input, 1, out) throws an exception.
  Actual: it doesn't.

[  FAILED  ] OpUnbindCopyIntOutTest.UnbindFailsWithWronglyAllocatedOutput (9 ms)
[----------] 1 test from OpUnbindCopyIntOutTest (9 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (9 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] OpUnbindCopyIntOutTest.UnbindFailsWithWronglyAllocatedOutput

 1 FAILED TEST

```
@eellison @digantdesai @GregoryComer  is this the same problem that  https://github.com/pytorch/pytorch/pull/130944#issuecomment-2418125356 is facing?

rec commented:
@clee2000 : Thanks for the test failure, it's quite revelatory!

I don't think it is the same problem as that other decomposition pull request though.

In this PR there is an actual change to the C++ implementation of `unbind_copy`, to fix [this issue](https://github.com/pytorch/pytorch/issues/130829).

After this change, `unbind_copy` accepts the same sorts of dimensions in `out=` arguments that other `_copy` ops do too: [this check](https://github.com/pytorch/pytorch/pull/134319/files#diff-e70cddcca22b07fef2f931799e771dd0f462dac8f35877861137db498643e6cfL4077), which is I believe what you are expecting, is gone, and replaced by [this series of checks](https://github.com/pytorch/pytorch/pull/134319/files#diff-e70cddcca22b07fef2f931799e771dd0f462dac8f35877861137db498643e6cfR4061), which are shared in common with another op, `split_with_sizes_copy_out` and also match the automatically generated Python unittests.

(I hasten to add that this API change mainly applies to edge cases: the code now fails on some cases where it succeeded but gave wrong answers before, and it succeeds in some cases it failed before. Code using this op that worked before will continue to work in exactly the same way it did before.)

So my theory is that that those two places where that executorch unit test fails are actually testing the old, incorrect behavior, and simply switching them to test that these two calls now actually work would be correct.

----

Assuming my theory is correct, the tricky part would be coordinating changing executorch's `kernels/test/op_unbind_copy_test.cpp` with merging this pytorch pull request again  

1. A primitive way to do this would be to have change ready to go in `executorch` and then when the breakage came in, to commit the change as a fix. That involves an intentional test breakage for a short period.

2. A more effortful way without breakage would be to temporarily rewrite the executorch test to handled both new and old cases explicitly catch and record exceptions, and test that _either_ both operations failed, _or_ both operations succeeded with expected values. After this decomposition pull request was merged, there'd be another change to that executorch test to get rid of the "exception" case.

rec committed:
Update

[ghstack-poisoned]

rec commented:
After the discussion on https://github.com/pytorch/pytorch/pull/130944, I decided to move `aten.unbind_copy.int` out of `_decomp._core_aten_decompositions_post_autograd()` and into `_inductor.decomposition.inductor_decompositions`.

The `unbind*` ops are known to be funky and this one wasn't working quite right for a long time without anyone noticing. Better to reduce the load on the downstream users of pytorch by making it an inductor-only decomposition!

rec committed:
Update

[ghstack-poisoned]

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your PR, but this is still failing the same test on ExecuTorch' -c ghfirst

I couldn't keep this merged while waiting for the fix because the long I wait, the higher the chance of getting a conflicts in github and fbcode.  To unblock this one, could you propose a PR to https://github.com/pytorch/executorch/blob/main/kernels/test/op_unbind_copy_test.cpp#L271 to update the test there?  I could help bring it to ExecuTorch team for their review and coordinate landing the two PRs.

Although, it's puzzling me that the test only fails internally, but not on PyTorch open source CI https://github.com/pytorch/pytorch/actions/runs/11481477468/job/31953571684#step:22:15616, so something is weird there, maybe that job is testing the wrong PyTorch version.

rec commented:
@huydhn: [Needs must when the devil drives!](https://en.wiktionary.org/wiki/needs_must_when_the_devil_drives)

But, oh, dear. If the same test succeeds in our CI and fails in yours, this is a problem. 🙁 

There are at least three possibilities:

1. `pytorch` isn't testing the right `torch` version with `executorch`
2. 'executorch` isn't testing the right `torch` version
3. There's some environmental difference between the two CI systems that leads to a different result

Case 1 is your theory and seems most likely. Case 2 is an obvious non-starter given the symptoms.

How can we prove that 1. is true (and then fix it)? Is there some quick and clever way to see it?

If this is true, it's worth expending some time to fix.

-----

As I pointed out in [a previous comment](https://github.com/pytorch/pytorch/pull/134319#issuecomment-2424953201), we don't have to have any rush or breakage in this process.

I would gladly write all three pull requests (excutorch/pytorch/executorch), but I am hampered by so far having failed to figure out how to run the executorch tests against a specific commit ID of pytorch - the installation instructions use a nightly build.

It's been several months since I tried this, I'm better at pytorch environments, I could go back to it and give it a whirl with my new knowledge: but surely someone must have done this before?  The executorch CI that caught this change must essentially be doing something like this...?

Any information you have on how to do this one tiny but blocking issue of running executorch tests against a specific commit ID of pytorch would be greatly appreciated!

huydhn commented:
So, there are pinned commits on both repos setting which commit to use:

* https://github.com/pytorch/pytorch/blob/main/.ci/docker/ci_commit_pins/executorch.txt
* and https://github.com/pytorch/executorch/blob/main/.ci/docker/ci_commit_pins/pytorch.txt

So, let's assume that we could let the test broken in trunk for a while before landing a forward fix.  The process would be:

1. Land https://github.com/pytorch/pytorch/pull/134319
2. Land a corresponding PR on https://github.com/pytorch/executorch
3. Update https://github.com/pytorch/pytorch/blob/main/.ci/docker/ci_commit_pins/executorch.txt commit from 2
4. Update https://github.com/pytorch/executorch/blob/main/.ci/docker/ci_commit_pins/pytorch.txt commit from 1

If you have 1 & 2 ready, I could help with 3 and 4

huydhn commented:
Also, there is a bug on ExecuTorch that wrongly uses nightly instead of the content of PyTorch from your PR, that explains why PR and trunk signals are all green.  The fix is on the way https://github.com/pytorch/executorch/pull/6564

huydhn commented:
@pytorchbot rebase -b main

rec committed:
Update

[ghstack-poisoned]

rec commented:
@pytorchbot rebase -b main

rec committed:
Update

[ghstack-poisoned]

rec committed:
Update

[ghstack-poisoned]

rec commented:
@pytorchbot rebase -b main

rec committed:
Update

[ghstack-poisoned]

rec commented:
@pytorchbot rebase

rec committed:
Update

[ghstack-poisoned]

rec commented:
@pytorchbot rebase

rec committed:
Update

[ghstack-poisoned]

rec commented:
@pytorchbot rebase -b main

rec committed:
Update

[ghstack-poisoned]

rec commented:
@pytorchbot rebase

rec committed:
Update

[ghstack-poisoned]

rec commented:
@huydhn: Hello, reviving this again.

I've updated this pull request a few times in the last month, and I'm still not seeing that test error that should be occurring, according to the above dialog.

I'd really love to get this pushed - it's the last part of a pull request I initially opened in June, and it fixes a user-observable issue.

Unfortunately, I have no way to even reproduce the breakage locally. This test on the pytorch side does not seem to break. 

And last time I tried this, the instructions to build executorch and its tests use pytorch nightly builds, and we weren't able to figure out how to get executorch to instead use a specific commit ID.

I don't want to make a pull request for executorch which I can't actually test against the problem.

Do you have any suggestions?

huydhn commented:
@pytorchbot merge

rec committed:
Update

[ghstack-poisoned]

rec commented:
Hey, @huydhn, thanks for doing this!

That was a weird error message, no idea what it is, I rebased against main and am trying again. Stay tuned!

rec commented:
@pytorchbot merge

rec committed:
Update

[ghstack-poisoned]

rec commented:
Sorry, I'm having trouble getting this commit working again after the rebase, I'll let you know...

rec committed:
Update

[ghstack-poisoned]

rec committed:
Update

[ghstack-poisoned]

rec commented:
@pytorchbot merge

rec committed:
Update

[ghstack-poisoned]

rec commented:
@pytorchbot merge

rec committed:
Update

[ghstack-poisoned]

rec commented:
@pytorchbot merge

rec commented:
@pytorchbot merge

rec commented:
@pytorchbot merge

rec commented:
@pytorchbot merge
